table/merging_iterator.cc

```diff
@@ -26,14 +26,92 @@
 #include "util/stop_watch.h"
 
 namespace ROCKSDB_NAMESPACE {
+// For merging iterator to process range tombstones, we treat the start and end
+// keys of a range tombstone as point keys and put them into the minHeap/maxHeap
+// used in merging iterator. Take minHeap for example, we are able to keep track
+// of currently "active" range tombstones (the ones whose start keys are popped
+// but end keys are still in the heap) in `active_`. This `active_` set of range
+// tombstones is then used to quickly determine whether the point key at heap
+// top is deleted (by heap property, the point key at heap top must be within
+// internal key range of active range tombstones).
+//
+// The HeapItem struct represents 3 types of elements in the minHeap/maxHeap:
+// point key and the start and end keys of a range tombstone.
+struct HeapItem {
+  HeapItem() = default;
+
+  enum Type { ITERATOR, DELETE_RANGE_START, DELETE_RANGE_END };
+  IteratorWrapper iter;
+  size_t level = 0;
+  std::string pinned_key;
+  // Will be overwritten before use, initialize here so compiler does not
+  // complain.
+  Type type = ITERATOR;
+
+  explicit HeapItem(size_t _level, InternalIteratorBase<Slice>* _iter)
+      : level(_level), type(Type::ITERATOR) {
+    iter.Set(_iter);
+  }
+
+  void SetTombstoneKey(ParsedInternalKey&& pik) {
+    pinned_key.clear();
+    // Range tombstone end key is exclusive. If a point internal key has the
+    // same user key and sequence number as the start or end key of a range
+    // tombstone, the order will be start < end key < internal key.
+    //
+    // Note: The op_type change of end key may not be necessary, but kept so
+    // that keys are non-decreasing in the case of a range tombstone and a user
+    // key straddles two SST files: note that we decrement end key's sequence
+    // number by 1 in TruncatedRangeDelIterator, which might make it the same
+    // sequence number as the smallest key in the next SST file.
+    ParsedInternalKey p(pik.user_key, pik.sequence, kValueTypeForSeek);
```

Comment:
Yes, I think it's reasonable, I can make that change. However, a similar block of code as this one is still needed in `merging_iter`. `smallest_` in TruncatedRangeDelIterator  can still be the same as a point key, and we need to modify it to have a smaller op_type so that it is always popped from the heap earlier than the point key. The end key of a un-truncated range tombstone can still be the same as a point key when, for example, it is ingested together with a key that is the same as the range tombstone end key. Also, I realize using `kValueTypeForSeek` is not enough: it is *equal* to the "highest-numbered ValueType". We will need to use `kValueTypeForSeek + 1` at least. I'll update the code to use `kTypeMaxValid`.

Reply:
If we are to consolidate the tombstone truncation logic, I think the following changes are needed in `TruncatedRangeDelIterator`:

- always sets `smallest_.op_type` to `kTypeMaxValid`
- sets `largest_.op_type` to `kTypeMaxValid` when `parsed_largest.sequence == kMaxSequenceNumber`

Reply:
Found out that `op_type` is needed for `largest_` even when `parsed_largest.sequence == kMaxSequenceNumber`. Without this, a range tombstone end key could be the same as a sentinel key from a level iterator, which makes their relative order not deterministic when popped from the heap.

Reply:
I updated the code to use `kTypeMaxValid` in `SetTombstoneKey()` to avoid the case when `kValueTypeForSeek` being equal to some valid value type. The `TruncatedRangeDelIterator` constructor is updated to adjust to `kValueTypeForSeek` when it decrements seqnum by 1. The op_type modifying logic is not moved to `TruncatedRangeDelIterator` since it would need to modify both `smallest_` and `largest_` op_type in all cases. I feel this logic is more specific to merging iterator, while `TruncatedRangeDelIterator` is also used in other scenarios, e.g. compaction code paths. But let me know if you think differently.

Reply:
Yes, file sentinel key can be processed before a range tombstone start key when scanning in reverse, even if the range tombstone is from the same file. This happens when they share the same user key and seqno. It requires special treatment when processing file sentinel keys while scanning in reverse: https://github.com/facebook/rocksdb/blob/010e0d25c8a9f8236919237f4484ba88a7f89e88/table/merging_iterator.cc#L988-L1004